Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIScriptingCocoa.mm

```diff
@@ -0,0 +1,188 @@
+    if (!hasPermission(webView.URL)) {
+        callback(std::nullopt);
+        return;
+    }
+
+    [webView _frames:makeBlockPtr([this, protectedThis = Ref { *this }, webView = RetainPtr { webView }, parameters = WTFMove(parameters), callback = WTFMove(callback)](_WKFrameTreeNode *mainFrame) mutable {
+        API::ContentWorld& executionWorld = *m_contentScriptWorld;
+        if (parameters.world == WebExtensionContentWorldType::Main)
+            executionWorld = API::ContentWorld::pageContentWorld();
+
+        Vector<RetainPtr<_WKFrameTreeNode>> frames;
+        if (parameters.allFrames || !parameters.frameIDs)
+            populateFrames(mainFrame, mainFrameID, parameters.allFrames, false, WTFMove(frames));
+        else
+            populateFrames(mainFrame, parameters.frameIDs.value(), WTFMove(frames));
+
+        size_t numberOfExecutions = frames.size();
+        Vector<WebExtensionScriptInjectionResultParameters> results;
+
+        for (auto& frame : frames) {
+            WKFrameInfo *info = frame.get().info;
+            NSURL *frameURL = info.request.URL;
+
+            if (!hasPermission(frameURL)) {
+                addResultAndCompleteIfDone(nil, info, toErrorString(@"scripting.executeScript()", nil, @"Extension does not have access to URL: %@", frameURL.absoluteString), numberOfExecutions, WTFMove(results), WTFMove(callback));
+                continue;
+            }
+
+            if (parameters.files) {
+                Vector<std::optional<SourcePair>> scriptPairs;
+                for (auto& file : parameters.files.value())
+                    scriptPairs.append(sourcePairForResource(file, m_extension));
+                
+                for (auto& script : scriptPairs) {
+                    [webView _evaluateJavaScript:script.value().first withSourceURL:script.value().second.value() inFrame:info inContentWorld:executionWorld.wrapper() completionHandler:makeBlockPtr([info = RetainPtr { info }, numberOfExecutions, results = WTFMove(results), callback = WTFMove(callback)](id resultOfJSExecution, NSError *error) mutable {
+                        addResultAndCompleteIfDone(resultOfJSExecution, info.get(), error.localizedDescription, numberOfExecutions, WTFMove(results), WTFMove(callback));
+                    }).get()];
+                }
+                
+                return;
+            }
+
+            NSArray *arguments = parameters.args ? createNSArray(parameters.args.value()).get() : @[ ];
+            NSString *function = [NSString stringWithFormat:@"return (%@)(...args)", (NSString *)(parameters.func.value())];
+
+            [webView _callAsyncJavaScript:function arguments:@{ @"args": arguments } inFrame:info inContentWorld:executionWorld.wrapper() completionHandler:makeBlockPtr([info = RetainPtr { info }, numberOfExecutions, results = WTFMove(results), callback = WTFMove(callback)](id resultOfJSExecution, NSError *error) mutable {
```

Comment:
Ditto: More moving of the `callback` and `results` variables for each step in the loop. Maybe `Arc` does something magic to keep these things alive somehow, but its confusing in the context of basic C++ operations like `WTFMove`.